pkg/hub/lease/controller.go

```diff
@@ -0,0 +1,180 @@
+		if err != nil {
+			return err
+		}
+		if updated {
+			syncCtx.Recorder().Eventf("ManagedClusterAvailableConditionUpdated", "update managed cluster %q available condition to %q, due to %q",
+				cluster.Name, availableCondition.Status, availableCondition.Reason)
+		}
+	}
+	return nil
+}
+
+// checkLease checks a given cluster's lease and determine the cluster whether is available.
+// 1. if the cluster does not have a lease, update the cluster available condition to unknown
+// 2. if the cluster have a lease and does not have an available condition, update the cluster
+//    available condition to unknown
+// 3. if the cluster have a lease and have an available condition
+//   - if the lease is constantly updated, keep the cluster available condition
+//   - if the lease is not constantly updated in a grace period, update the cluster available
+//     condition to unknown
+func (c *leaseController) checkLease(cluster *clusterv1.ManagedCluster, gracePeriod time.Duration) (*clusterv1.StatusCondition, error) {
+	clusterLeaseName := fmt.Sprintf("cluter-%s-lease", cluster.Name)
+	observedLease, err := c.leaseLister.Leases(cluster.Name).Get(clusterLeaseName)
+	if errors.IsNotFound(err) {
+		return &clusterv1.StatusCondition{
+			Type:    clusterv1.ManagedClusterConditionAvailable,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "ManagedClusterLeaseNeverUpdated",
+			Message: "Registration agent never updated its lease.",
+		}, nil
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	observedCond := helpers.FindManagedClusterCondition(cluster.Status.Conditions, clusterv1.ManagedClusterConditionAvailable)
+	if observedCond == nil {
+		return &clusterv1.StatusCondition{
+			Type:    clusterv1.ManagedClusterConditionAvailable,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "ManagedClusterAvailableStatusNeverUpdated",
+			Message: "Registration agent never posted its available status.",
+		}, nil
+	}
+
+	lastClusterLease := c.clusterLeaseMap.get(clusterLeaseName)
+	// update the hub lease probe time and last lease if the managed cluster lease is constantly updated.
+	// in next, we will use hub probe time to determine the managed cluster lease whether is constantly
+	// updated in a grace period to avoid the problem of time synchronization between hub and managed
+	// cluster.
+	if lastClusterLease.lease == nil || lastClusterLease.lease.Spec.RenewTime.Before(observedLease.Spec.RenewTime) {
+		lastClusterLease.lease = observedLease
+		lastClusterLease.probeTimestamp = metav1.Now()
+		c.clusterLeaseMap.set(clusterLeaseName, lastClusterLease)
+	}
+
+	if metav1.Now().After(lastClusterLease.probeTimestamp.Add(gracePeriod)) {
+		return &clusterv1.StatusCondition{
+			Type:    clusterv1.ManagedClusterConditionAvailable,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "ManagedClusterAvailable",
+			Message: "Registration agent stopped updating its lease.",
```

Comment:
indicate how long it has been since the lease was updated.

Reply:
added